src/lib.rs

```diff
@@ -15,16 +15,27 @@ pub use ser::Serializer;
 
 type Result<T> = std::result::Result<T, Error>;
 
-fn static_str_to_js(s: &'static str) -> JsString {
+// This API isn't great, but there isn't a good way for us to return a reference;
+// it would have to have a static lifetime, and while the cache will live forever,
+// it has to reallocate as it grows, invalidating any references to it.
+//
+// The reason we want to use a reference in the first place is to avoid having to
+// repeatedly clone and destroy these strings in the `JsValue` heap, which takes time.
+fn with_js_string<T, F>(s: &'static str, f: F) -> T
+where
+    F: FnOnce(&JsString) -> T,
+{
     thread_local! {
-        static CACHE: std::cell::RefCell<fnv::FnvHashMap<&'static str, JsString>> = Default::default();
+        // Since we're mainly optimising for converting the exact same string literal over and over again,
+        // which will always have the same pointer, we can speed things up by indexing by the string's pointer
+        // instead of its value.
+        static CACHE: std::cell::RefCell<fnv::FnvHashMap<*const u8, JsString>> = Default::default();
```

Comment:
This one is somewhat problematic, because even if this optimisation doesn't exist today, technically linker is allowed to merge strings to save size. Especially so in Rust where nul terminator doesn't exist.

For example, strings "abc" and "abcde" could result in a same location in the binary & the same pointer to a string "abcde", just used with different lengths. Clashing of those pointers as cache keys would then cause subtle and unpleasant bugs.

To future-proof code against such optimisations, the only safe option would be to use the fat pointer as a key, which then means you need to always cache 16 bytes. Since this cache is mostly intended for field names, many of those would be actually shorter, so, instead, this would be a slowdown in many cases.

At least I remember trying `*const str` at some point in the past and not seeing any obvious wins to justify it. You can recheck if you want though. But I don't think we should go with just `*const u8` given the risks.

Reply:
Good catch! `*const str` still seems to be a bit quicker, so I'll go with that.